### PR TITLE
Zero pad chunks before saving

### PR DESCRIPTION
**Pull request by schoentoon**
This way the world is better compressable and backups will be smaller. I made an external tool that does this after the fact and it can make backups of the entire world around 5 to 20% smaller in my testings.

### CODE REVIEW

File: Spigot-Server-Patches/0555-Zero-pad-chunks-before-saving-them-to-the-disk.patch
Comment (aikar) [inline Spigot-Server-Patches/0555-Zero-pad-chunks-before-saving-them-to-the-disk.patch]: this is better served by having a single reused buffer of the entire block size, calculate the last block the "partial" data writes into, and then zero out that block and the remainder BEFORE writing the real data, so we can avoid allocating buffers every write.

This static zero buffer should also be a Direct allocated byte buffer too.

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0555-Zero-pad-chunks-before-saving-them-to-the-disk.patch
Comment (aikar) [inline Spigot-Server-Patches/0555-Zero-pad-chunks-before-saving-them-to-the-disk.patch]: This needs to be a Direct Memory buffer, and static, as otherwise the JVM copies memory every single write

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0555-Zero-pad-chunks-before-saving-them-to-the-disk.patch
Comment (Shevchik) [reply Spigot-Server-Patches/0555-Zero-pad-chunks-before-saving-them-to-the-disk.patch]: This should also use & 4095 instead. Mod operator is slower than bit and.

File: Spigot-Server-Patches/0592-Zero-pad-chunks-before-saving-them-to-the-disk.patch
Comment (schoentoon) [inline Spigot-Server-Patches/0592-Zero-pad-chunks-before-saving-them-to-the-disk.patch]: I would like to point out that I changed this purely so the `bytebuffer.limit(size)` later on wouldn't throw an exception. As it will throw an exception if the new size is beyond the capacity according to https://docs.oracle.com/javase/8/docs/api/java/nio/Buffer.html#limit-int-
I found that the capacity with the original value would go from 8096 (so too small) to 16192 etc. 8192 probably makes more sense anyway, as it's a multiple of the chunk size of 4096.

Comment (aikar) [reply Spigot-Server-Patches/0592-Zero-pad-chunks-before-saving-them-to-the-disk.patch]: I feel like that's a bug on mojang anyways where they originally had 4096 and wanted to double it and just changed the 4 to an 8 instead.

**Review by aikar — DISMISSED** [review]
Approving but dev team please test this before merge.

File: Spigot-Server-Patches/0592-Zero-pad-chunks-before-saving-them-to-the-disk.patch
Comment (hugmanrique) [inline Spigot-Server-Patches/0592-Zero-pad-chunks-before-saving-them-to-the-disk.patch]: If `count` is a multiple of 4096, this yields an incorrect value (+4096 of what's desired).
We need a "`Math.ceilDiv`", which doesn't exist. However, you can prove ceilDiv(m, n)=-floorDiv(-m, n). Maybe Mojang already has a util for this?

Alternatively, albeit slower, you can use `((int) Math.ceil(count / 4096D)) * 4096`

**Review by hugmanrique — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0592-Zero-pad-chunks-before-saving-them-to-the-disk.patch
Comment (schoentoon) [reply Spigot-Server-Patches/0592-Zero-pad-chunks-before-saving-them-to-the-disk.patch]: Seems like I incorrectly copied that from one of the other functions. Update incoming where I just use said function instead.

Comment (Spottedleaf) [reply Spigot-Server-Patches/0592-Zero-pad-chunks-before-saving-them-to-the-disk.patch]: int size = (this.count + 4095) & ~4095

**Comment by schoentoon** [comment]
Made the requested changes. Simply bumping because github doesn't seem to give out notifications for resolving conversations and/or force pushes to existing pull requests :)

**Comment by MiniDigger** [comment]
@limbo-app ready

(just marking it as ready so that this shows up in my filter, haven't touched this yet)

**Review by limbo-app — DISMISSED** [review]
Pull request automatically approved for merging - a core team member has previously approved this pull request.

**Comment by MiniDigger** [comment]
@limbo-app ready

**Review by limbo-app — APPROVED** [review]
Pull request automatically approved for merging - a core team member has previously approved this pull request.

**Comment by MiniDigger** [comment]
sorry for taking so long, I rebased and tested this locally 3 times but never finished properly...
thanks for your contribution!

**Pull request merged**
